src/rp2_common/hardware_i2c/i2c.c

```diff
@@ -75,6 +75,13 @@ uint i2c_set_baudrate(i2c_inst_t *i2c, uint baudrate) {
     invalid_params_if(I2C, hcnt < 8);
     invalid_params_if(I2C, lcnt < 8);
 
+    // The SDA hold time should be at least 300 ns, per the I2C standard.
+    // sda_hold_count [cycles] = freq_in [cycles/s] * 300ns * (1s / 1e9ns)
+    // Reduce 300/1e9 to 3/1e7 to avoid numbers that don't fit in uint.
+    // Add 1 to avoid division truncation.
+    uint sda_hold_count = ((freq_in * 3) / 10000000) + 1;
+    invalid_params_if(I2C, sda_hold_count > lcnt - 2);
```

Comment:
Is the usage of the `invalid_params_if` macro appropriate here, given that both `sda_hold_count` and `lcnt` are calculated values? :man_shrugging:  Perhaps a regular `assert` would be better?

Reply:
`invalid_params_if` was used on line 83 because it's also used [a few lines up on line 76](https://github.com/raspberrypi/pico-sdk/blob/2ab09b28feb9fdedd784a16269b69646deea1562/src/rp2_common/hardware_i2c/i2c.c#L76):
```c
    invalid_params_if(I2C, lcnt < 8);
```
 If `invalid_params_if` is appropriate for `lcnt < 8` I assumed it's appropriate for `sda_hold_count > lcnt - 2`.

Is this assumption incorrect?

That being said, I'm not super familiar with the pico-sdk codebase so I've no objections to making this change.

Reply:
I'm not sure either, which is why I shrugged. It's really a question for @kilograham

Reply:
invalid_params_if to catch the result of out of range parameters (unless the function returns an error code which most don't) - can be for a knock on calculation from the original 

assert() should be used to check internal consistency/assumptions

Reply:
Ok, I think I get it now.

Switched from `invalid_params_if` to `assert` for checking the consistency of the following constraint mentioned at the top of page 512 in the RP2040 datasheet:

> The programmed SDA hold time during transmit (IC_SDA_TX_HOLD) cannot exceed at any time the duration of the low part of scl. Therefore the programmed value cannot be larger than N_SCL_LOW-2, where N_SCL_LOW is the duration of the low part of the scl period measured in ic_clk cycles.